web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/AnalyzeIncomingEntities.php

```diff
@@ -0,0 +1,432 @@
+      $category = $this->getCategory($provenance_record);
+      $this->setSolutionCategory($solution_id, $category);
+    }
+  }
+
+  /**
+   * Computes the category of a solution given its provenance activity record.
+   *
+   * @param \Drupal\rdf_entity\RdfInterface $activity
+   *   The solution provenance activity.
+   *
+   * @return string
+   *   The category ID.
+   */
+  protected function getCategory(RdfInterface $activity): string {
+    $collection_id = $this->getPipeline()->getCollection();
+
+    // If the provenance activity record is new, there was no previous attempt
+    // to federate this solution.
+    if ($activity->isNew()) {
+      return 'not_federated';
+    }
+    // If the solution is already associated with another collection, we can't
+    // federate it in the scope of this pipeline's collection.
+    elseif ($activity->get('provenance_associated_with')->value !== $collection_id) {
+      return 'invalid_collection';
+    }
+    // If there is an existing provenance activity enabled record, this incoming
+    // entity has been previously federated.
+    elseif ($activity->get('provenance_enabled')->value) {
+      // In case the entity is marked as federated, check all entities related
+      // to the entity. If none of them changed, mark it as unchanged.
+      $entity_list = $this->getSolutionsWithDependenciesAsFlatList([$activity->get('provenance_entity')->value]);
+      if (!$this->isSolutionChanged($entity_list)) {
+        return 'federated_unchanged';
+      }
+
+      return 'federated';
+    }
+    // Otherwise this solution has been previously blacklisted.
+    return 'blacklisted';
+  }
+
+  /**
+   * Returns whether a solution has changed given a list of entity ids.
+   *
+   * The solution will be considered as changed if one of the following occur
+   * for at least one of the entries:
+   * - The entity does not have a changed property. This means we cannot know if
+   *   it has changed.
+   * - The provenance record of the entity is new. This meanst the entity has
+   *   not been federated before.
+   * - The entity's changed time is more recent than the provenance record's
+   *   provenance_started property.
+   *
+   * @param string[] $entity_ids
+   *   A list of entity ids that the solution is related to.
+   *
+   * @return bool
+   *   Whether the solution or one of its related entities have changed since
+   *   the last import.
+   */
+  protected function isSolutionChanged(array $entity_ids): bool {
+    /** @var \Drupal\rdf_entity\RdfInterface[] $solutions */
+    $solutions = $this->getRdfStorage()->loadMultiple($entity_ids, ['staging']);
+    $provenance_records = $this->provenanceHelper->loadOrCreateEntitiesActivity($entity_ids);
+
+    foreach ($solutions as $id => $solution) {
+      // Licences are stored in Joinup so changes do not affect incoming
+      // solution.
+      if ($solution->bundle() === 'licence') {
+        continue;
+      }
```

Comment:
We are importing a solution and all related entities to a solution. However, we only select the solution on the UserSelection step. Due to that last part, a solution is changed if one of the following cases:
- The solution itself has changed.
- One of the related entities to the solution has changed.
Because, if one of the related entities changed, we need to import the whole solution tree (the solution and its dependencies) so that the changed entities are re-imported.

The PhpDoc is written properly and explains that this is not only solution ids. 
However, there is a problem with the name of the variable here. So I am renaming this.